### PR TITLE
Add support to network v2 🐡

### LINKED ISSUE #278: Composefile v2: support networks
Add support for managing networks for v2 services (meaning for projects that have `NetworkConfig` populated).

### PR DESCRIPTION
**Pull request by vdemeester**
This PR brings the first and main step to add support for networking for v2 compose files 🐹.

- Create a `docker/network` package, with testing
- Add some yml to try it out
- Update some stuff in container and service. Note though there is a lot of FIXME(s)/TODO(s) because some of the things added will be refactored/moved with service.go and container.go refactoring.

There is probably still some weird edge cases where it does not works correctly but it does the job for the main use cases. What is still to do :

- [x] Complete unit tests, depending on docker/engine-api#238 to *mock* `NotFound` network errors and thus test the call to `NetworkCreate`.
- [x] Add some integration tests
- [x] Support `service:…` and `container:…` network mode :wink: 

As said before, I made some changes *against my will* because of the way `Container` and `Service` interact with the current code base ; they will be taken care of when refactoring them to make `Container` a little bit less intelligent (and way less *talkative* with the API) and `Service` a little bit more :angel:. 

/cc @joshwget @ibuildthecloud @shouze

Closes #278 

🐸

*Note: this builds on top of #309 and #310 to make integration test pass :angel:*

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

### CODE REVIEW

**Comment by thaJeztah** [comment]
Awesome job!

File: project/project.go [hunk 1]
```diff
@@ -28,24 +28,26 @@ type Project struct {
...
+	context         *Context
+	reload          []string
+	upCount         int
+	listeners       []chan<- events.Event
+	hasListeners    bool
 }
 
 // NewProject creates a new project with the specified context.
-func NewProject(context *Context, runtime RuntimeProject, parseOptions *config.ParseOptions) *Project {
+func NewProject(context *Context, runtime RuntimeProject, networksFactory NetworksFactory, parseOptions *config.ParseOptions) *Project {
```
Comment (joshwget) [inline project/project.go]: It seems a little odd that `serviceFactory` is in `context` but `networksFactory` isn't. Not sure which of the two places is better, but we should probably be consistent.

Comment (vdemeester) [inline project/project.go]: @joshwget that's true. I'll put it in `Context` for now, and we'll see later :+1:

File: docker/container.go [hunk 2]
```diff
@@ -379,26 +381,12 @@ func holdHijackedConnection(tty bool, inputStream io.ReadCloser, outputStream, e
 	return nil
 }
 
-// Up creates and start the container based on the image name and send an event
-// to notify the container has been created. If the container exists but is stopped
-// it tries to start it.
-func (c *Container) Up(ctx context.Context, imageName string) error {
```
Comment (vdemeester) [inline docker/container.go]: This did not make sense at all (a container do not `Up` itself) and the change was easy enough to make in the PR (it even make the code path easier to follow :angel: ).

File: docker/service.go [hunk 3]
```diff
@@ -282,7 +282,13 @@ func (s *Service) Info(ctx context.Context, qFlag bool) (project.InfoSet, error)
...
 // Start implements Service.Start. It tries to start a container without creating it.
 func (s *Service) Start(ctx context.Context) error {
-	return s.up(ctx, "", false, options.Up{})
+	return s.collectContainersAndDo(ctx, func(c *Container) error {
+		if err := s.connectContainerToNetworks(ctx, c); err != nil {
+			return err
+		}
+		return c.Start(ctx)
+	})
+	// return s.up(ctx, "", false, options.Up{})
```
Comment (vdemeester) [inline docker/service.go]: Oh… I should remove this…

File: docker/service.go [hunk 4]
```diff
@@ -282,7 +282,13 @@ func (s *Service) Info(ctx context.Context, qFlag bool) (project.InfoSet, error)
 
 // Start implements Service.Start. It tries to start a container without creating it.
 func (s *Service) Start(ctx context.Context) error {
-	return s.up(ctx, "", false, options.Up{})
```
Comment (vdemeester) [inline docker/service.go]: Calling `up` here would create the container if not already created. Which is not right, as said, `Start` start the container without creating it.

File: project/project.go [hunk 5]
```diff
@@ -183,11 +180,45 @@ func (p *Project) load(file string, bytes []byte) error {
...
 		}
+		networkNames = append(networkNames, name)
+	}
+
+	for name, config := range serviceConfigs {
+		// Handle NetworkMode
+		// FIXME(vdemeester) not sure it's enough, this might need some rework (here, in service and container)
+		if config.NetworkMode == "" {
+			if p.isNetworkEnabled() {
+				config.NetworkMode = networkNames[0]
```
Comment (joshwget) [inline project/project.go]: I'm a little confused with these couple of lines. It looks like a default `NetworkMode` for v2 services is being set based on `networkNames[0]`, but shouldn't `networkNames` be empty if there aren't any network configurations?

Comment (vdemeester) [inline project/project.go]: @joshwget oh you are right, I should check if `networkNames` is empty before doing that :sweat:

File: docker/network/network_test.go [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+		expectedError         string
+	}{
+		{
+			network: &Network{
+				name:     "net1",
+				external: true,
+			},
+			inspectError:  fmt.Errorf("any error"),
+			expectedError: "any error",
+			// expectedError: "Error: No such network: net1",
```
Comment (joshwget) [inline docker/network/network_test.go]: Need this still?

Comment (vdemeester) [inline docker/network/network_test.go]: yes, this is the part where i depend on a change in engine-api to finish unit testing 😇

Comment (joshwget) [inline docker/network/network_test.go]: Just checking :smile:

**Comment by joshwget** [comment]
Other than the one comment, everything within `project` looks good to me! I think an `EmptyNetworks` struct similar to `EmptyService` would be handy, but I can add that in a follow up PR if you'd like.

I'm not very familiar with Docker networking, so I can't really review the `docker` package changes all that well. Integration tests would definitely be nice :smile:

File: docker/network/network_test.go (hunk 6 above)
Comment (vdemeester) [inline docker/network/network_test.go]: yay engine-api PR merged, i'll do a revendor (separate) and will update this one

**Comment by vdemeester** [comment]
Ok unit tests are completed, tried it out on some networks v2 compose files, the only thing that is not implemented is the new `networkMode` options (`container:mycontainer`, `service:myservice`)..

And still need some integration tests.

**Comment by vdemeester** [comment]
To not repeat stuff, I did not write integration test (`CliSuite`), and I'm using the *network* part of the acceptance test to validate the behaviour.

To run all *network* acceptance tests, you can do the following

```bash
$ ACCEPTANCE_TEST_ARGS="-k network" make test-acceptance
# […]
============================================================== test session starts ==============================================================
platform linux2 -- Python 2.7.11 -- py-1.4.31 -- pytest-2.7.2 -- /home/vincent/lib/go/src/github.com/docker/libcompose/venv/bin/python2.7
rootdir: /home/vincent/lib/go/src/github.com/docker/libcompose/venv/compose, inifile: tox.ini
plugins: cov
collected 108 items
tests/acceptance/cli_test.py::CLITestCase::test_config_external_network Running process: 7653
FAILED
tests/acceptance/cli_test.py::CLITestCase::test_run_detached_connects_to_network <- tests/integration/testcases.py Running process: 7657
Running process: 7665
Killing networks_app_1 ... done
Killing networks_db_1 ... done
Killing networks_web_1 ... done
Removing networks_app_1 ... done
Removing networks_db_1 ... done
Removing networks_web_1 ... done
PASSED
tests/acceptance/cli_test.py::CLITestCase::test_run_interactive_connects_to_network <- tests/integration/testcases.py Running process: 7679
Running process: 7687
Running process: 7695
Killing networks_app_1 ... done
Killing networks_web_1 ... done
Killing networks_db_1 ... done
Removing networks_app_1 ... done
Removing networks_web_1 ... done
Removing networks_db_1 ... done
PASSED
tests/acceptance/cli_test.py::CLITestCase::test_up_external_networks <- tests/integration/testcases.py Running process: 7709
Running process: 7716
Killing networks_web_1 ... done
Removing networks_web_1 ... done
PASSED
tests/acceptance/cli_test.py::CLITestCase::test_up_missing_network <- tests/integration/testcases.py Running process: 7726
FAILED
tests/acceptance/cli_test.py::CLITestCase::test_up_with_default_network_config <- tests/integration/testcases.py Running process: 7731
Killing networks_another_1 ... done
Killing networks_simple_1 ... done
Removing networks_another_1 ... done
Removing networks_simple_1 ... done
PASSED
tests/acceptance/cli_test.py::CLITestCase::test_up_with_external_default_network <- tests/integration/testcases.py Running process: 7742
Running process: 7749
Killing networks_simple_1 ... done
Killing networks_another_1 ... done
Removing networks_simple_1 ... done
Removing networks_another_1 ... done
PASSED
tests/acceptance/cli_test.py::CLITestCase::test_up_with_network_aliases <- tests/integration/testcases.py Running process: 7760
Killing networks_web_1 ... done
Removing networks_web_1 ... done
PASSED
tests/acceptance/cli_test.py::CLITestCase::test_up_with_network_mode <- tests/integration/testcases.py Running process: 7769
Stderr: time="2016-06-23T19:58:34+02:00" level=info msg="[0/5] [bridge]: Starting "
time="2016-06-23T19:58:34+02:00" level=info msg="[0/5] [host]: Starting "
time="2016-06-23T19:58:34+02:00" level=info msg="[0/5] [container]: Starting "
time="2016-06-23T19:58:34+02:00" level=info msg="[0/5] [none]: Starting "
time="2016-06-23T19:58:34+02:00" level=info msg="[0/5] [service]: Starting "
time="2016-06-23T19:58:34+02:00" level=error msg="Failed Starting service : Error response from daemon: network service:bridge not found"
time="2016-06-23T19:58:34+02:00" level=info msg="[1/5] [host]: Started "
time="2016-06-23T19:58:35+02:00" level=info msg="[2/5] [container]: Started "
time="2016-06-23T19:58:35+02:00" level=info msg="[3/5] [none]: Started "
time="2016-06-23T19:58:35+02:00" level=error msg="Failed to start: service : Error response from daemon: network service:bridge not found"
time="2016-06-23T19:58:35+02:00" level=info msg="[4/5] [bridge]: Started "
Error response from daemon: network service:bridge not found
Stdout: Creating network "networks_default" with driver ""
Killing networks_container_1 ... done
Killing networks_bridge_1 ... done
Killing networks_none_1 ... done
Killing networks_host_1 ... done
Removing networks_container_1 ... done
Removing networks_bridge_1 ... done
Removing networks_none_1 ... done
Removing networks_host_1 ... done
FAILED
tests/acceptance/cli_test.py::CLITestCase::test_up_with_network_static_addresses <- tests/integration/testcases.py Running process: 7784
Killing networks_web_1 ... done
Removing networks_web_1 ... done
PASSED
tests/acceptance/cli_test.py::CLITestCase::test_up_with_networks <- tests/integration/testcases.py Running process: 7794
Killing networks_app_1 ... done
Killing networks_db_1 ... done
Killing networks_web_1 ... done
Removing networks_app_1 ... done
Removing networks_db_1 ... done
Removing networks_web_1 ... done
PASSED
=================================================================== FAILURES ====================================================================
___________________________________________________ CLITestCase.test_config_external_network ____________________________________________________
tests/acceptance/cli_test.py:231: in test_config_external_network
    assert 'networks' in json_result
E   AssertionError: assert 'networks' in "No help topic for 'config'"
______________________________________________________ CLITestCase.test_up_missing_network ______________________________________________________
tests/integration/testcases.py:50: in wrapper
    return f(self, *args, **kwargs)
tests/acceptance/cli_test.py:627: in test_up_missing_network
    assert 'Service "web" uses an undefined network "foo"' in result.stderr
E   AssertionError: assert 'Service "web" uses an undefined network "foo"' in 'time="2016-06-23T19:58:31+02:00" level=fatal msg="Failed to read project: Service \"web\" uses an undefined network \"foo\"" \n'
E    +  where 'time="2016-06-23T19:58:31+02:00" level=fatal msg="Failed to read project: Service \"web\" uses an undefined network \"foo\"" \n' = ProcessResult(stdout=u'', stderr=u'time="2016-06-23T19:58:31+02:00" level=fata...ed to read project: Service \\"web\\" uses an undefined network \\"foo\\"" \n').stderr
_____________________________________________________ CLITestCase.test_up_with_network_mode _____________________________________________________
tests/integration/testcases.py:50: in wrapper
    return f(self, *args, **kwargs)
tests/acceptance/cli_test.py:641: in test_up_with_network_mode
    self.dispatch(['-f', filename, 'up', '-d'], None)
tests/acceptance/cli_test.py:129: in dispatch
    return wait_on_process(proc, returncode=returncode)
tests/acceptance/cli_test.py:50: in wait_on_process
    assert proc.returncode == returncode
E   AssertionError: assert 1 == 0
E    +  where 1 = <subprocess.Popen object at 0x7f40a5910510>.returncode
====================================================== 97 tests deselected by '-knetwork' =======================================================
============================================== 3 failed, 8 passed, 97 deselected in 12.42 seconds ===============================================
# […]
```

Right now, one tests fails because of `logrus` (will be handled a little later), one because the config file is not implemented yet (see other PR) and the final one on `network-mode` because this PR do not implement it… I'm thinking of implementing it in a follow-up PR (because this one is big enough…).

**Comment by vdemeester** [comment]
All right @joshwget gonna merge this and we'll iterate over this in case of bug or anything else :wink:

File: docker/convert.go [hunk 7]
```diff
@@ -114,7 +117,7 @@ func ports(c *config.ServiceConfig) (map[nat.Port]struct{}, nat.PortMap, error)
 }
 
 // Convert converts a service configuration to an docker API structures (Config and HostConfig)
-func Convert(c *config.ServiceConfig, ctx project.Context) (*container.Config, *container.HostConfig, error) {
+func Convert(c *config.ServiceConfig, ctx project.Context, clientFactory composeclient.Factory) (*container.Config, *container.HostConfig, error) {
```
Comment (joshwget) [inline docker/convert.go]: Is there any way we can get around having to use the Docker daemon here? The process of converting a `ServiceConfig` to a container `Config` and `HostConfig` is really useful even in use cases without a Docker daemon. We're pretty dependent on this for Rancher Compose.

If not, maybe we can just split this into two functions, one that doesn't depend on Docker and another that does (while reusing the former).

I was actually thinking about moving the conversion process to a separate package to make it explicitly independent of the `docker` package (which I see as more responsible for the actual interactions with the Docker daemon).

File: project/project.go [hunk 8]
```diff
@@ -164,30 +166,89 @@ func (p *Project) load(file string, bytes []byte) error {
...
 		}
+
+		p.networks = networks
 	}
 
 	return nil
 }
 
+func (p *Project) isNetworkEnabled() bool {
+	return p.configVersion == "2"
```
Comment (joshwget) [inline project/project.go]: If no networks are specified, it looks like there's a slight difference in behavior between v1 and v2.
- v1 will behave exactly the same as before this PR and make no network API calls
- v2 will create a create a `default` network

Is this correct?

Comment (vdemeester) [inline project/project.go]: Yes, this is correct :angel:

File: docker/convert.go (hunk 7 above)
Comment (vdemeester) [inline docker/convert.go]: Yeah, I do agree, having `client.Factory` here is a pain. I think we can refactor some code to get this out of here but this will be easier with the `Container` refactoring :angel:.

I like the idea of making the conversion process in a separate package :wink:.

Comment (joshwget) [inline docker/convert.go]: Great! Glad to head the conversion process isn't fundamentally tied to using a Docker daemon :smile:

Comment (vdemeester) [inline docker/convert.go]: I'll look into it tomorrow :wink:

Comment (vdemeester) [inline docker/convert.go]: The `client.Factory` is there for handling `container:xxxx` network modes (and in fact, just get the container ID :smile:). Handling networkMode should be done elsewhere (but I ran out of idea in this PR :smile_cat: ).
